docs - ALTER/CREATE/DROP TABLE, CTAS sql ref page updates

lisakowen opened the pull request:
bring ALTER/CREATE/DROP TABLE, CREATE TABLE AS sql ref pages up to postgres v12 content.  there is a bunch of greenplum-specific info on the pages.

in this PR:
- the changes to ALTER/CREATE TABLE are VERY extensive, not sure if i captured everything ...
- i didn't change much of the existing greenplum-specific content on the classic partitioning syntax.
- CREATE TABLE - the postgres v12 page lists a bunch of storage_parameters (https://www.postgresql.org/docs/12/sql-createtable.html) - not clear to me if greenplum supports any of those that we haven't documented.

i tagged greenplum-specific info that i want to verify with XXX.  ALTER/CREATE TABLE pages have the bulk of these.  is the content still true in v7?

questions:
1.  am i distinguishing the modern/classic partitioning syntax in the most appropriately way?  the synopses presents the modern syntax first, and then calls out the classic partitioning syntax later.  i tried to do the same with the parameter descriptions.
2.  since greenplum does not support referential integrity (foreign key) constraints, should we just remove this syntax from the ref pages, rather than adding notes/words that greenplum accepts it but does not enforce it?
3.  some new clauses are introduced, like EXCLUDE constraints. it is not clear to me if there are any differences in support between classic vs modern in this and other new clauses.

doc review site links (behind vpn); you can navigate to all ref pages from here:  https://docs-staging.vmware.com/en/draft/VMware-Greenplum/sqlref-table/greenplum-database/GUID-ref_guide-sql_commands-CREATE_TABLE.html

huansong commented:
Not yet looking into the changes. For the questions:

> am i distinguishing the modern/classic partitioning syntax in the most appropriately way? the synopses presents the modern syntax first, and then calls out the classic partitioning syntax later. i tried to do the same with the parameter descriptions.

I think that's fine. In that way we regard the modern syntax as the "default" which is good. When we call out the classic syntax we may provide a link to https://docs-staging.vmware.com/en/draft/VMware-Tanzu-Greenplum/part2/greenplum-database/GUID-admin_guide-ddl-ddl-partition.html#choosing-the-partitioning-syntax which gives details to anyone wondering what classic syntax is (or https://docs-staging.vmware.com/en/draft/VMware-Tanzu-Greenplum/part2/greenplum-database/GUID-admin_guide-ddl-about-part-changes.html)

> since greenplum does not support referential integrity (foreign key) constraints, should we just remove this syntax from the ref pages, rather than adding notes/words that greenplum accepts it but does not enforce it?

I think we don't need to do that. Consider someone who just wants to search information about referential integrity on this page: we should let them see the notes instead of seeing nothing. For anyone who isn't interested in referential integrity, they'll just skip the notes ( I think few people is going to read this page paragraph by paragraph, most people are just searching for specific stuff).

> some new clauses are introduced, like EXCLUDE constraints. it is not clear to me if there are any differences in support between classic vs modern in this and other new clauses.

Interesting. But I think partitioned table doesn't support EXCLUDE constraints: https://www.postgresql.org/docs/12/sql-createtable.html#:~:text=Partitioned%20tables%20do%20not%20support%20EXCLUDE%20constraints . So classic vs modern shouldn't matter. But not sure about other new clauses, LMK if there's anything that we should verify.

mperezfuster commented:
Thanks so much for such extensive work Lisa.
Note I am reviewing this without much (or any) background regarding classic vs modern partitioning syntax.
I agree with Huangson in that it could be very useful to point to other non-reference pages that provide details on the new partitioning. However on the pages he points to I am unable to find the terms classic/modern. I can see uniform/flexible syntax. Are these equivalent? I am very sorry if I am saying something stupid.
Regarding the Compatibility section, I have the feeling that the previous syntax looked a bit tidier and more compact. I know we get the right side navigation from turning the items into titles, but they look a bit too big to me and not as much as a list as they used to, they look a bit harder to read. But it's only my opinion.

lisakowen commented:
hey @mperezfuster .  thank you for taking a look.  the links that huansong provided are for the review site for a previous PR.  i did not update the review site after the final edits.  check out the official doc page here:  https://docs.vmware.com/en/VMware-Greenplum/7/greenplum-database/admin_guide-ddl-ddl-partition.html.  (note that updates/examples for modern syntax are in-progress, and not yet reflected in this topic.)

huansong reviewed (commented):
Went through the `CREATE TABLE` page. Will look into `ALTER TABLE ` next.

lisakowen commented on gpdb-doc/markdown/ref_guide/sql_commands/CREATE_TABLE.html.md:
```diff
@@ -197,364 +191,464 @@ and subpartition\_element is:
+USING access\_method
+:   The optional `USING` clause specifies the table access method to use to store the contents for the new table you are creating; the method must be an access method of type `TABLE`. Set to `heap` to access the table as a heap-storage table, `ao_row` to access the table as an append-optimized table with row-oriented storage (AO), or `ao_column` to access the table as an append-optimized table wi ...
 
-  <p class="note">
+:   <p class="note">
 <strong>Note:</strong>
 Although you can specify the table's access method using <code>WITH (appendoptimized=true|false, orientation=row|column)</code> VMware recommends that you use <code>USING <access_method></code> instead.
 </p>
   
-WITH \( storage\_parameter=value \)
-:   The `WITH` clause can specify storage parameters for tables, and for indexes associated with a `UNIQUE` or `PRIMARY` constraint. See [Storage Parameters](#storage-parameters), below, for details.  
+WITH ( storage\_parameter=value )
+:   The `WITH` clause specifies optional storage parameters for a table or index; see [Storage Parameters](#storage_parameters) below for details. For backward-compatibility the `WITH` clause for a table can also include `OIDS=FALSE` to specify that rows of the new table should not contain OIDs (object identifiers), `OIDS=TRUE. is no longer supported.
 
 ON COMMIT
-:   The behavior of temporary tables at the end of a transaction block can be controlled using `ON COMMIT`. The three options are:
+:   You can control the behavior of temporary tables at the end of a transaction block using `ON COMMIT`. The three options are:
 
 :   **PRESERVE ROWS** - No special action is taken at the ends of transactions for temporary tables. This is the default behavior.
 
-:   **DELETE ROWS** - All rows in the temporary table will be deleted at the end of each transaction block. Essentially, an automatic `TRUNCATE` is done at each commit.
+:   **DELETE ROWS** - All rows in the temporary table will be deleted at the end of each transaction block. Essentially, Greenplum Database performs an automatic [TRUNCATE](TRUNCATE.html) at each commit. When used on a partitioned table, this operation is not cascaded to its partitions.
 
-:   **DROP** - The temporary table will be dropped at the end of the current transaction block.
+:   **DROP** - The temporary table will be dropped at the end of the current transaction block. When used on a partitioned table, this action drops its partitions and when used on tables with inheritance children, it drops the dependent children.
 
 TABLESPACE tablespace
-:   The name of the tablespace in which the new table is to be created. If not specified, the database's default tablespace is used, or [temp\_tablespaces](../config_params/guc-list.html) if the table is temporary.
+:   The name of the tablespace in which the new table is to be created. If not specified, the database's  [default\_tablespace](../config_params/guc-list.html#default_tablespace) is consulted, or [temp\_tablespaces](../config_params/guc-list.html) if the table is temporary. For partitioned tables, since no storage is required for the table itself, the tablespace specified overrides `default_table ...
 
 USING INDEX TABLESPACE tablespace
-:   This clause allows selection of the tablespace in which the index associated with a `UNIQUE` or `PRIMARY KEY` constraint will be created. If not specified, the database's default tablespace is used, or [temp\_tablespaces](../config_params/guc-list.html) if the table is temporary.
+:   This clause allows selection of the tablespace in which the index associated with a `UNIQUE`, `PRIMARY KEY`, or `EXCLUDE` constraint will be created. If not specified, the database's  [default\_tablespace](../config_params/guc-list.html#default_tablespace) is used, or [temp\_tablespaces](../config_params/guc-list.html) if the table is temporary.
 
-DISTRIBUTED BY \(column \[opclass\], \[ ... \] \)
+DISTRIBUTED BY \( column \[opclass\] \[, ... \] \)
 DISTRIBUTED RANDOMLY
 DISTRIBUTED REPLICATED
-:   Used to declare the Greenplum Database distribution policy for the table. `DISTRIBUTED BY` uses hash distribution with one or more columns declared as the distribution key. For the most even data distribution, the distribution key should be the primary key of the table or a unique column \(or set of columns\). If that is not possible, then you may choose `DISTRIBUTED RANDOMLY`, which will sen ...
+:   Used to declare the Greenplum Database distribution policy for the table. `DISTRIBUTED BY` uses hash distribution with one or more columns declared as the distribution key. For the most even data distribution, the distribution key should be the primary key of the table or a unique column (or set of columns). If that is not possible, then you may choose `DISTRIBUTED RANDOMLY`, which will send  ...
```
thanks for catching this merge error on my part.

lisakowen commented on gpdb-doc/markdown/ref_guide/sql_commands/CREATE_TABLE.html.md:
```diff
@@ -197,364 +191,464 @@ and subpartition\_element is:
+:   For a column, the optional `ENCODING` clause specifies the type of compression and block size for the column data. Valid column storage\_directives are `compresstype`, `compresslevel`, and `blocksize`.
 
-:   The clause is valid only for append-optimized, column-oriented tables.
+:   This clause is valid only for append-optimized, column-oriented (AO/CO) tables.
```
removed it.

lisakowen commented on gpdb-doc/markdown/ref_guide/sql_commands/CREATE_TABLE.html.md:
```diff
@@ -197,364 +191,464 @@ and subpartition\_element is:
 ### <a id="storage_parameters"></a>Storage Parameters
 
-The `WITH` clause can specify storage parameters for tables, and for indexes associated with a `UNIQUE` or `PRIMARY` constraint. Storage parameters for indexes are documented on the [CREATE INDEX](CREATE_INDEX.html.md) reference page. Note that you can also set storage parameters for a particular partition or subpartition by declaring the `WITH` clause in the partition specification. The lowest-l ...
+The `WITH` clause can specify storage parameters for tables, and for indexes associated with a `UNIQUE`, `PRIMARY KEY`, or `EXCLUDE` constraint. Storage parameters for indexes are documented on the [CREATE INDEX](CREATE_INDEX.html.md) reference page. The storage parameters currently available for tables are listed below. For many of these parameters, as shown, there is an additional parameter wit ...
```
are you saying we need to add toast.vacuum_index_cleanup because vacuum_index_cleanup is in our list?  or are you saying we need to add more or all of the parameters listed on the postgres 12 CREATE TABLE ref page https://www.postgresql.org/docs/12/sql-createtable.html?  how would i go about identifying which ones greenplum supports?

huansong replied:
The second - we need to add all of those new upstream parameters. GP7 pulled all of them and they should all be supported for heap tables.

lisakowen replied:
thank you for reviewing, @huansong!

i added all of the storage parameters from the postgresql 12 section except for user_catalog_table.  a few notes/questions:
1.  i wasn't sure if the calculations for toast_tuple_target should be adjusted because greenplum uses a different default blocksize.  (or maybe they are referring to a different blocksize?)
2. we do not document all of the related gucs.  where we have docs, i included an xref.  those without docs are just code font (green).  do i need to add these into the greenplum docs guc reference?

huansong replied:
Thanks for adding them! I'll review them in details and get back to your questions.

huansong replied:
1. Yes the default value `toast_tuple_target` will be adjusted in GPDB corresponding to a bigger block size and I believe it would be 8184 instead of 2040. Good catch!
2. Yes we should add those missing GUCs. But note that, many of the per-table options are not really effective, I'll review them one-by-one and comment accordingly. (thanks to input from @bmdoil and @soumyadeep2007 ).

huansong commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -281,49 +412,40 @@ new\_name
 
 table\_constraint
-:   New table constraint for the table. Note that foreign key constraints are currently not supported in Greenplum Database. Also a table is only allowed one unique constraint and the uniqueness must be within the Greenplum Database distribution key.
+:   New table constraint for the table. Note that foreign key constraints are currently not supported in Greenplum Database. XXX Also a table is only allowed one unique constraint and the uniqueness must be within the Greenplum Database distribution key. XXX
```
`XXX` comment is still true

huansong commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -409,19 +561,29 @@ SPLIT PARTITION
+The key word COLUMN is noise and can be omitted.
 
-Greenplum Database does not currently support foreign key constraints. For a unique constraint to be enforced in Greenplum Database, the table must be hash-distributed \(not `DISTRIBUTED RANDOMLY`\), and all of the distribution key columns must be the same as the initial columns of the unique constraint columns.
+When a column is added with `ADD COLUMN`, all existing rows in the table are initialized with the column's default value, or `NULL` if no `DEFAULT` clause is specified. Adding a column with a non-null default or changing the type of an existing column will require the entire table and indexes to be rewritten. As an exception, if the `USING` clause does not change the column contents and the old t ...
```
Just a note for future: this needs revisit when we work the table rewrite aspect for `ADD COLUMN`.

huansong commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -409,19 +561,29 @@ SPLIT PARTITION
+
+To force immediate reclamation of space occupied by a dropped column, you can run one of the forms of `ALTER TABLE` that performs a rewrite of the whole table. This results in reconstructing each row with the dropped column replaced by a null value.
+
+XXX
```
Note for future: the table below will need to be modified to:

```
Operation (See Note)	Append-Optimized, Column-Oriented	Append-Optimized	Heap
ALTER COLUMN TYPE	No	Yes	Yes
ADD COLUMN	No	No	No
ALTER COLUMN SET ENCODING	No	N/A	N/A
```

huansong commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -430,53 +592,91 @@ This table lists the `ALTER TABLE` operations that require a table rewrite when
-> **Note** Dropping a system `oid` column also requires a table rewrite.
+XXX
 
-When a column is added with `ADD COLUMN`, all existing rows in the table are initialized with the column's default value, or `NULL` if no `DEFAULT` clause is specified. Adding a column with a non-null default or changing the type of an existing column will require the entire table and indexes to be rewritten. As an exception, if the `USING` clause does not change the column contents and the old t ...
+> **Important** The forms of `ALTER TABLE` that perform a table rewrite are not MVCC-safe. After a table rewrite, the table will appear empty to concurrent transactions if they are using a snapshot taken before the rewrite occurred. See [MVCC Caveats](https://www.postgresql.org/docs/12/mvcc-caveats.html) for more details.
 
-> **Important** The forms of `ALTER TABLE` that perform a table rewrite on an append-optimized table are not MVCC-safe. After a table rewrite, the table will appear empty to concurrent transactions if they are using a snapshot taken before the rewrite occurred. See [MVCC Caveats](https://www.postgresql.org/docs/12/mvcc-caveats.html) for more details.
+XXX Take special care when altering or dropping columns that are part of the Greenplum Database distribution key as this can change the distribution policy for the table.
```
`XXX` comment is still true

huansong commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -430,53 +592,91 @@ This table lists the `ALTER TABLE` operations that require a table rewrite when
 
--   Adding or dropping a column.
--   Changing the data type of column.
+- The table name specified in the `ALTER TABLE` command cannot be the name of a partition within a table.
```
This statement is very confusing. Maybe better:
```
The table name specified in the ALTER TABLE command has to be the actual table name of the partition, not the partition alias that is specified in the classic syntax.
```

huansong commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -430,53 +592,91 @@ This table lists the `ALTER TABLE` operations that require a table rewrite when
+    -   Adding or dropping constraints.
+    -   Splitting an external partition.
+
+XXX
```
Other `XXX` comments look fine to me

huansong reviewed (commented):
Went through the `ALTER TABLE` page.

lisakowen commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -23,65 +23,137 @@ ALTER TABLE [IF EXISTS] <name>
+  SET TABLESPACE <new_tablespace>
+  SET { LOGGED | UNLOGGED }
+  SET ( <storage_parameter> [= <value>] [, ...])
+  SET WITH (<set_with_parameter> = <value> [, ...])
```
thank you for providing these examples!

lisakowen commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -141,135 +207,200 @@ and subpartition\_element is:
+:   Changes the owner of the table, sequence, view, materialized view, or foreign table to the specified user.
+
+RENAME
+:   Changes the name of a table (or an index, sequence, view, materialized view, or foreign table), the name of an individual column in a table, or the name of a constraint of the table. When renaming a constraint that has an underlying index, the index is renamed as well. There is no effect on the stored data. XXX Note that Greenplum Database does not permit renaming distribution key columns. XX ...
```
the 7th occurrence of rename on https://docs.vmware.com/en/VMware-Greenplum/6/greenplum-database/ref_guide-sql_commands-ALTER_TABLE.html includes a variation of the statement about not allowing rename of distribution keys.

lisakowen commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -141,135 +207,200 @@ and subpartition\_element is:
+:   `SET STATISTICS` acquires a `SHARE UPDATE EXCLUSIVE` lock.
 
-    **RESET \( attribute\_option \[, ...\] \)**— Sets or resets per-attribute options. Currently, the only defined per-attribute options are `n_distinct` and `n_distinct_inherited`, which override the number-of-distinct-values estimates made by subsequent [ANALYZE](ANALYZE.html) operations. `n_distinct` affects the statistics for the table itself, while `n_distinct_inherited` affects the statisti ...
+SET ( attribute\_option = value [, ... ] )
+RESET ( attribute\_option [, ...] )
+:   Sets or resets per-attribute options. Currently, the only defined per-attribute options are `n_distinct` and `n_distinct_inherited`, which override the number-of-distinct-values estimates made by subsequent [ANALYZE](ANALYZE.html) operations. `n_distinct` affects the statistics for the table itself, while `n_distinct_inherited` affects the statistics gathered for the table plus its inheritanc ...
```
added a SET ENCODING section.  also added a notice to use SET ENCODING in the case you describe.

lisakowen commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -337,8 +459,38 @@ new\_tablespace
+DISTRIBUTED RANDOMLY
+ DISTRIBUTED REPLICATED
+:   Specifies the distribution policy for a table. Changing a hash distribution policy causes the table data to be physically redistributed, which can be resource intensive. If you declare the same hash distribution policy or change from hash to random distribution, data will not be redistributed unless you declare `SET WITH (reorganize=true)`.
+:   Changing to or from a replicated distribution policy always causes Greenplum Database to redistribute the table data.
```
i removed this block and combined the description with SET DISTRIBUTED.

lisakowen commented on gpdb-doc/markdown/ref_guide/sql_commands/ALTER_TABLE.html.md:
```diff
@@ -141,135 +207,200 @@ and subpartition\_element is:
+
+You must own the table to use `ALTER TABLE`. To change the schema or tablespace of a table, you must also have `CREATE` privilege on the new schema or tablespace. To add the table as a new child of a parent table, you must own the parent table as well. Also, to attach a table as a new partition of the table, you must own the table being attached. To alter the owner, you must also be a direct or i ...
+
+> **Note** XXX Memory usage increases significantly when a table has many partitions, if a table has compression, or if the blocksize for a table is large. If the number of relations associated with the table is large, this condition can force an operation on the table to use more memory. For example, if the table is a CO table and has a large number of columns, each column is a relation. An oper ...
```
i made the example more generic.

huansong commented on gpdb-doc/markdown/ref_guide/sql_commands/CREATE_TABLE.html.md:
```diff
@@ -197,364 +191,464 @@ and subpartition\_element is:
+You cannot define a table with more than 1600 columns. (In practice, the effective limit is usually lower because of tuple-length constraints.)
 
-    In Greenplum Database, values of type `CHAR(n)` are padded with trailing spaces to the specified width n. The values are stored and displayed with the spaces. However, the padding spaces are treated as semantically insignificant. When the values are distributed, the trailing spaces are disregarded. The trailing spaces are also treated as semantically insignificant when comparing two values of ...
+XXX
```
The comments are all good.

nit: `DECLARE...FOR UPDATE, and triggers are not supported with append-optimized tables. CLUSTER on append-optimized tables is only supported over B-tree indexes.` should belong to two separate paragraphs.

huansong reviewed (commented):
Went over the rest.

lisakowen commented:
@huansong , thank you for the very thorough review.  i believe i have addressed all of your comments.  see https://github.com/greenplum-db/gpdb/pull/15707#discussion_r1244345432 for a few remaining questions.

huansong commented on gpdb-doc/markdown/ref_guide/sql_commands/CREATE_TABLE.html.md:
```diff
@@ -197,364 +184,521 @@ and subpartition\_element is:
+vacuum_truncate, toast.vacuum_truncate (boolean)
+:   Enables or disables vacuum to attempt to truncate any empty pages at the end of this table. The default value is `true`. If `true`, `VACUUM` and autovacuum truncate the empty pages, and the disk space for the truncated pages is returned to the operating system. Note that the truncation requires an `ACCESS EXCLUSIVE` lock on the table. The `TRUNCATE` parameter of `VACUUM`, if specified, overri ...
+
+autovacuum_vacuum_threshold, toast.autovacuum_vacuum_threshold (integer)
+:   Per-table value for the [autovacuum_vacuum_threshold](../config_params/guc-list.html#autovacuum_vacuum_threshold) server configuration parameter.
```
Per-table `autovacuum_vacuum_threshold` currently is not effective: (1) we do not allow autovacuum user-table, so change it on user tables doesn't have an impact; (2) we allow autovacuum for catalogs but then user cannot (and probably shouldn't) change `autovacuum_vacuum_threshold` of a catalog.

So let's point that out here in the doc: user can set them but they are not effective currently.

Same situation is for these options below (basically just except for the ones for analyze)
```
autovacuum_vacuum_scale_factor
autovacuum_vacuum_cost_delay
autovacuum_vacuum_cost_limit
autovacuum_freeze_min_age
autovacuum_freeze_max_age
autovacuum_freeze_table_age
autovacuum_multixact_freeze_min_age
autovacuum_multixact_freeze_max_age
autovacuum_multixact_freeze_table_age
log_autovacuum_min_duration
```

P.S. we have a story to allow change these settings for the auxiliary tables of AO tables. After that we will need to make doc changes here accordingly. But for now we can ignore that.

lisakowen replied:
thank you @huansong .  i've added a general note to the storage options section, and related/specific notes to the description of the impacted options.  let me know if you don't think the wording is appropriate.

re: adding the missing options to the server configuration parameter (guc) reference, i will do that in a separate PR if ok with you.

huansong replied:
Thanks! Yes separate PR for those would be fine

huansong reviewed (approved):
LGTM!
